File: StarcUp.Core/Src/Business/Units/Types/UpgradeFrames.cs
namespace StarcUp.Business.Units.Types
{
    /// <summary>
    /// 업그레이드별 소요 프레임 수
    /// </summary>
    public enum UpgradeFrames : ushort
    {
        // Terran Armor & Weapons (레벨별 다름: 1lv=4000, 2lv=4480, 3lv=4960)
        Terran_Infantry_Armor = 4000,
        Terran_Vehicle_Plating = 4000,
        Terran_Ship_Plating = 4000,
        Terran_Infantry_Weapons = 4000,
        Terran_Vehicle_Weapons = 4000,
        Terran_Ship_Weapons = 4000,

        // Zerg Armor & Weapons (레벨별 다름: 1lv=4000, 2lv=4480, 3lv=4960)
        Zerg_Carapace = 4000,
        Zerg_Flyer_Carapace = 4000,
        Zerg_Melee_Attacks = 4000,
        Zerg_Missile_Attacks = 4000,
        Zerg_Flyer_Attacks = 4000,

        // Protoss Armor & Weapons (레벨별 다름: 1lv=4000, 2lv=4480, 3lv=4960)
        Protoss_Ground_Armor = 4000,
        Protoss_Air_Armor = 4000,
        Protoss_Ground_Weapons = 4000,
        Protoss_Air_Weapons = 4000,
        Protoss_Plasma_Shields = 4000,

        // Terran Unit Upgrades
        U_238_Shells = 1500,
        Ion_Thrusters = 1500,
        Titan_Reactor = 2500,
        Ocular_Implants = 2500,
        Moebius_Reactor = 2500,
        Apollo_Reactor = 2500,
        Colossus_Reactor = 2500,
        Caduceus_Reactor = 2500,
        Charon_Boosters = 2000,

        // Zerg Unit Upgrades
        Ventral_Sacs = 2400,
        Antennae = 2000,
        Pneumatized_Carapace = 2000,
        Metabolic_Boost = 1500,
        Adrenal_Glands = 1500,
        Muscular_Augments = 1500,
        Grooved_Spines = 1500,
        Gamete_Meiosis = 2500,
        Metasynaptic_Node = 2500,
        Chitinous_Plating = 2000,
        Anabolic_Synthesis = 2000,

        // Protoss Unit Upgrades
        Singularity_Charge = 2500,
        Leg_Enhancements = 2000,
        Scarab_Damage = 2500,
        Reaver_Capacity = 2500,
        Gravitic_Drive = 2500,
        Sensor_Array = 2000,
        Gravitic_Boosters = 2000,
        Khaydarin_Amulet = 2500,
        Apial_Sensors = 2000,
        Gravitic_Thrusters = 2500,
        Carrier_Capacity = 1500,
        Khaydarin_Core = 2500,
        Argus_Jewel = 2500,
        Argus_Talisman = 2500,

        // Special values
        Upgrade_60 = 0,
        None = 0,
        Unknown = 0,
        MAX = 0
    }
}

File: StarcUp.UI/src/types/game/upgradeFrames.ts
import { UpgradeType } from './UpgradeType';

// 업그레이드별 소요 프레임 수 및 최대 레벨
export interface UpgradeInfo {
  frames: number[];  // 레벨별 프레임 수 배열 (1lv, 2lv, 3lv)
  maxLevel: number;
}

// Armor & Weapons 업그레이드 (3레벨)
const ARMOR_WEAPON_FRAMES = [4000, 4480, 4960];

export const upgradeFrames: Record<UpgradeType, UpgradeInfo> = {
  // Terran Armor & Weapons (3레벨)
  [UpgradeType.Terran_Infantry_Armor]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Terran_Vehicle_Plating]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Terran_Ship_Plating]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Terran_Infantry_Weapons]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Terran_Vehicle_Weapons]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Terran_Ship_Weapons]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },

  // Zerg Armor & Weapons (3레벨)
  [UpgradeType.Zerg_Carapace]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Zerg_Flyer_Carapace]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Zerg_Melee_Attacks]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Zerg_Missile_Attacks]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Zerg_Flyer_Attacks]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },

  // Protoss Armor & Weapons (3레벨)
  [UpgradeType.Protoss_Ground_Armor]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Protoss_Air_Armor]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Protoss_Ground_Weapons]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Protoss_Air_Weapons]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },
  [UpgradeType.Protoss_Plasma_Shields]: { frames: ARMOR_WEAPON_FRAMES, maxLevel: 3 },

  // Terran Unit Upgrades (1레벨)
  [UpgradeType.U_238_Shells]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Ion_Thrusters]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Titan_Reactor]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Ocular_Implants]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Moebius_Reactor]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Apollo_Reactor]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Colossus_Reactor]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Caduceus_Reactor]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Charon_Boosters]: { frames: [2000], maxLevel: 1 },

  // Zerg Unit Upgrades (1레벨)
  [UpgradeType.Ventral_Sacs]: { frames: [2400], maxLevel: 1 },
  [UpgradeType.Antennae]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Pneumatized_Carapace]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Metabolic_Boost]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Adrenal_Glands]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Muscular_Augments]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Grooved_Spines]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Gamete_Meiosis]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Metasynaptic_Node]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Chitinous_Plating]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Anabolic_Synthesis]: { frames: [2000], maxLevel: 1 },

  // Protoss Unit Upgrades (1레벨)
  [UpgradeType.Singularity_Charge]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Leg_Enhancements]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Scarab_Damage]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Reaver_Capacity]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Gravitic_Drive]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Sensor_Array]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Gravitic_Boosters]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Khaydarin_Amulet]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Apial_Sensors]: { frames: [2000], maxLevel: 1 },
  [UpgradeType.Gravitic_Thrusters]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Carrier_Capacity]: { frames: [1500], maxLevel: 1 },
  [UpgradeType.Khaydarin_Core]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Argus_Jewel]: { frames: [2500], maxLevel: 1 },
  [UpgradeType.Argus_Talisman]: { frames: [2500], maxLevel: 1 },

  // Special values
  [UpgradeType.Upgrade_60]: { frames: [0], maxLevel: 0 },
  [UpgradeType.None]: { frames: [0], maxLevel: 0 },
  [UpgradeType.Unknown]: { frames: [0], maxLevel: 0 },
};

// 업그레이드 정보 가져오기 헬퍼 함수
export function getUpgradeFrameInfo(upgradeType: UpgradeType): UpgradeInfo {
  return upgradeFrames[upgradeType] || { frames: [0], maxLevel: 0 };
}

// 특정 레벨의 프레임 수 가져오기
export function getUpgradeFramesForLevel(upgradeType: UpgradeType, level: number): number {
  const info = getUpgradeFrameInfo(upgradeType);
  if (level < 1 || level > info.maxLevel) return 0;
  return info.frames[level - 1] || 0;
}